accepted/2023/simplify-output-paths.md

```diff
@@ -0,0 +1,130 @@
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>\publish` - Publish output with `RuntimeIdentifier`
+
+## Proposed behavior
+
+We will introduce a new output path format, called the "artifacts" output format after the default folder name it will use.  By convention, the artifacts folder will be located at the solution or repo level rather within each project folder.
+
+The recommended way to opt in to the new output path format will be to set the `UseArtifactsOutput` property to `true` in a `Directory.Build.props` file.
+
+The artifacts folder will by default be named `.artifacts`.  The convention we will use to determine which folder the artifacts folder should go in will be:
+
+- If a `Directory.Build.props` file was used in the build, put the artifacts folder in the same folder where `Directory.Build.props` was found, which will typically be at the root of a repo.  Since MSBuild already probes for `Directory.Build.props`, we should be able to use this folder for the artifacts with minimal perf impact
+- Look for a folder with a `.sln` file in it, walking up the directory tree from the current project folder to the root.  When a `.sln` file is found, put the artifacts folder in that same directory.  We may need an additional intrinsic MSBuild function to do this, as the `GetDirectoryNameOfFileAbove` function doesn't support wildcards for the file name
+- If neither a `Directory.Build.props` file nor a `.sln` file was found, then put the artifacts folder in the project folder.
+
+To override the default artifacts folder path, the `ArtifactsPath` property can be used.  If `ArtifactsPath` is set, then `UseArtifactsOutput` will default to `true` and it won't be necessary to set both properties.  `ArtifactsPath` should normally be set in a `Directory.Build.props` file.  For example:
+
+```xml
+<Project>
+    <PropertyGroup>
+        <ArtifactsPath>$(MSBuildThisFileDirectory)\output</ArtifactsPath>
+    </PropertyGroup>
+</Project>
+```
+
+Under the artifacts folder will be nested folders for the type of the output, the project name, and the pivots, ie `<ArtifactsPath>\<Type of Output>\<Project Name>\<Pivots>`.
+
+- Type of Output - This will be a value such as `bin`, `publish`, `intermediates`, or `package`.  Projects will be able to override this value, for example to separate shipping and non-shipping artifacts into different folders
+  - `bin` will be the folder for the normal build output.  Projects can override this with `ArtifactsOutputName`
+  - `publish` will be the folder for the publish output.  Projects can override this with `ArtifactsPublishOutputName`
+  - `package` will be the folder for the package output, where .nupkg folders are placed. (Or should it be `packages`?  Feedback welcome.)  Projects can override this with `ArtifactsPackageOutputName`
+  - `intermediates` will be the folder for the intermediate output.  (Or should we stick with `obj`?  Feedback welcome.)
```

Comment:
I know there's been lots of feedback and probably discussions around this, but I think it'd be nice to just go back to obj. Given that it's only a name change unlike the other directories, it probably makes the chnages simpler to understand.

Reply:
I agree, and I think that both bin and obj should be in a folder called `cache` or `caches`. This is what gradle does in java which is the closest equivalent I have used.
```
<RepositoryRoot>/
└── <ArtifactsPath>/
    ├── caches/
    │   ├── bin
    │   ├── obj
    │   └── packages
    ├── publish
    └── package
```

Reply:
@ChristopherHaws Can you help me understand more about how caching would work with these folders?

From the [GitHub documentation](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows) you linked, it seems like the cache is used for inputs to the build, such as NuGet packages that are downloaded.  It doesn't seem like it would be helpful for the `bin`, `obj`, and `packages` folders here, as they are outputs of the build and will likely be changing for each build.

Also /cc @Nirmal4G who had some feedback about caching.

Reply:
My thought was that these files could be cached so that the build server could do an incremental build instead of a cold build.

Reply:
What would you use for the cache key?  MSBuild incremental builds are based on file timestamps, do you know if the caching preserves those?

Reply:
Hmm, for some reason I remember there being a file in `obj` that contained the hashes of all the files in the build, but I can't seem to find it now. Is that something that used to exist and now doesn't or am I just going crazy? 😆

Reply:
> My thought was that these files could be cached so that the build server could do an incremental build instead of a cold build.

Which build server are you referring to here?

Reply:
I was talking about caching NuGet restores on Containerized builds but it could also be applicable on full build caching but I don't know why anyone would want to do a full build cache.

Reply:
I was talking about caching NuGet restores on Containerized builds but it could also be applicable on full build caching but I don't know why anyone would want to do a full build cache.

This is also why I've isolated NuGet restores into `ext` folder separate from `obj`.